src/content/blog/astro-4100.mdx

```diff
@@ -0,0 +1,184 @@
+lang: "en"
+---
+
+import BlogSummary from "/src/components/BlogSummary.astro";
+
+<BlogSummary>
+  Astro 4.10 is out with experimental type-safe environment variables
+</BlogSummary>
+
+Full release highlights include:
+
+- [**Experimental: astro:env module**](#experimental-astro-env)
+- [**Rewrite for all HTTP methods**](#rewrite-for-all-http-methods)
+- [**Embedding Astro in server frameworks**](#embedding-astro)
+- [**Renderer helper functions**](#renderer-helper-functions)
+
+To upgrade an existing project, use the automated `@astrojs/upgrade` CLI tool. Alternatively, upgrade manually by running the upgrade command for your package manager:
+
+```sh
+# Recommended:
+npx @astrojs/upgrade
+
+# Manual:
+npm install astro@latest
+pnpm upgrade astro --latest
+yarn upgrade astro --latest
+```
+
+## Experimental: astro:env
+
+Astro 4.10 introduces a new experimental built-in module, `astro:env`.
+
+Environment variables allow you to configure your app with different values in different environments. But this comes with a great deal of complexity:
+
+- Some variables are used in the client and some only on the server.
+- Server variables are often __secrets__, things like API keys that you absolutely do not want exposed in the client, and you also do not want inlined into the server build either (which can be viewed by anyone with access to the build output).
+- Some variables are __required__ for your app to function at all; whereas other are optional enhancements.
+- Runtimes such as Cloudflare and Deno have different APIs for reading variables, creating a dev/prod difference you need to deal with.
+
+We built `astro:env` to provide more control and structure over environment variables. Here's how you might define a schema for them in your config:
+
+```js
+import { defineConfig, envField } from 'astro/config';
+
+export default defineConfig({
+  env: {
+    schema: {
+      DATABASE_URL: envField.string({
+        context: 'client',
+        access: 'public'
+      }),
+      API_PORT: envField.number({
+        context: 'server',
+        access: 'secret',
+        default: 7000
+      })
+    }
+  }
+})
+```
+
+Once defined you can use your variables by importing them from the `astro:env` module like so:
+
+```js
+import { API_PORT } from 'astro:env/server';
+
+// API_PORT is a number
+// ...
+```
+
+Some times you still want to read a variable that is not defined in the schema. This might be because the variable name is dynamic itself such as `BACK_SERVER_1` and `BACK_SERVER_2`, for example. Or it might be because you are reading variables inside of an Astro integration and don't want to impose the schema onto the app. In these cases `getSecret` is available to read runtime variables and works with any runtime (Cloudflare, Node.js, Deno).
+
+```js
+import { getSecret } from 'astro:env/server';
+
+function getServerEndpoint(num: number) {
+  return getSecret('BACKUP_SERVER_' + num);
+}
+```
+
```

Comment:
Feels a bit weird to ask 😅 but is it possible to get my name in there, like Ming-jun Lu in 4.6 or Oliver in 4.4?

Reply:
Yes absolutely! Did so here: https://github.com/withastro/astro.build/pull/1116/commits/dc36fb307d0112a4f9f7291b7b951957fceb81d1